### ISSUE TITLE
Refresh access token doesn't work

### ISSUE THREAD

**Issue by AntonyARHS** [aws-amplify/amplify-android#2380]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-android/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-android/discussions).

### Language and Async Model

Kotlin

### Amplify Categories

Authentication

### Gradle script dependencies

<details>
    implementation 'com.amplifyframework:core-kotlin:2.5.0'
    implementation 'com.amplifyframework:aws-auth-cognito:2.6.0'
</details>

### Environment information

<details>

```
# Put output below this line


```

</details>

### I'm using the following process in my app: (fresh install)

- Amplify.Auth.signIn -> The user correctly sign in. 
- Once the authentication is done, I use Amplify.Hub.subscribe(HubChannel.AUTH) to check all the events. 
- When the user is loggedin, I used "Amplify.Auth.fetchUserAttributes()" to get the user information.

### To get the tokens, I'm using: 

"Amplify.Auth.fetchAuthSession" before each call to our API. 

### The configuration about the user pool we're using: 

Acces token + ID token validity = 7 minutes
Refresh token = 3 hours

### The problem:

When the app is killed / Restarted, we are trying to use "fetchUserAttributes" and we have the exception : "NotAuthorizedException{message=Failed since user is not authorized., cause=NotAuthorizedException(message=Access Token has expired), recoverySuggestion=Check whether the given values are correct and the user is authorized to perform the operation.} 
" which is expected actually, because it was after 7 minutes. 

**BUT**, when calling "Amplify.Auth.fetchAuthSession" to get a new access token usable, we're still getting the previous one... Impossible to get a new one until we force a logout / login, which is not expected at all. We expect to be able to receive a new access token using a refresh token still valid...

I saw the first time the tokens were generated, everything was "correct" for the IAT / Exp properties. Then, I expect to be able with Amplify to have a new access token automatically when calling "fetchAuthSession". 

How to proceed?

**Comment by ranshine** [comment]
I am facing same issue with amplify android library with  1.38.3v, killing the app and after some time if i open the app Amplify.Auth.fetchAuthSession is not retrieving tokens.

**Comment by sdhuka** [comment]
@AntonyARHS @ranshine 

We are not able to repro the issue using the cognito expiry settings you mentioned. Could you please provide us logs by adding android logging plugin during amplify configuration? **_please make sure to redact any sensitive information._**

`Amplify.addPlugin(AndroidLoggingPlugin(LogLevel.VERBOSE))`

**Comment by tylerjroach** [comment]
@AntonyARHS 

I see that you have a short lifespan for your refresh token (3 hrs). I'd like to clarify that refresh token age is the maximum age of the token. Upon new calls to refresh user pool tokens, the access/id tokens update, but the refresh token does not. This means that no login in the application will last longer than 3 hrs without having to re-sign in.

Is it possible that when you have closed the app and reopened, the refresh token may have also expired so that refreshing id/access token is no longer possible?

If you are signing in and seeing the NotAuthorizedException for the access token within the first 3 hours, please let us know and provide the logging details @sdhuka requested. `fetchUserAttributes` calls `fetchAuthSession` internally and should initiate the access token refresh. If it is failing, we should be able to see it in the logs.

@ranshine 
Amplify v1 and v2 are very different implementations. Please open up a new ticket with details if you are experiencing issues. Also, make sure your refresh token validity duration, seeing the note about refresh tokens being the maximum life of a the User Pool credentials.

**Comment by AntonyARHS** [comment]
Yes, I know, but I had this problem with a refreshtoken that should still be valid. 

I'll let you know when I have the bug again. It's a bit annoying to debug this kind of thing when you have to wait 7 minutes to try to reproduce it :)

**Comment by adamtester** [comment]
Hello, just for visibility a few of us are having a similar issue over on flutter, but only reproducible on Android:

- https://github.com/aws-amplify/amplify-flutter/issues/3057
- https://github.com/aws-amplify/amplify-flutter/issues/186

**Comment by Sypotu** [comment]
I followed the [getting started guide](https://docs.amplify.aws/lib/auth/getting-started/q/platform/android/) and could easily replicate this issue:

these are my Cognito client settings:
![image](https://github.com/aws-amplify/amplify-android/assets/9402291/2f786791-2502-40f2-ac7d-df1015b1aa27)


And here is my code, trying to refresh the tokens once the id token is expired and the refresh token is not yet expired:
```
implementation 'com.amplifyframework:core:2.8.4'
implementation 'com.amplifyframework:aws-auth-cognito:2.8.4'
```

```java
Amplify.addPlugin(new AWSCognitoAuthPlugin());
Amplify.addPlugin(new AndroidLoggingPlugin(LogLevel.VERBOSE));
```

```java
@Override
    protected void onResume() {
        super.onResume();

        Amplify.Auth.signIn(
                "username",
                "password",
                result -> Log.i("AuthQuickstart", result.isSignedIn() ? "Sign in succeeded" : "Sign in not complete"),
                error -> Log.e("AuthQuickstart", error.toString())
        );

        refreshAndLogTokens();

        try {
            Thread.sleep(8 * 60 * 1000);   // wait for 8min, the idToken will expire
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        refreshAndLogTokens();

    }
    private void refreshAndLogTokens()
    {
        AuthFetchSessionOptions options = AuthFetchSessionOptions.builder().forceRefresh(true).build();
        Amplify.Auth.fetchAuthSession(options,
                result -> {
                    AWSCognitoAuthSession cognitoAuthSession = (AWSCognitoAuthSession) result;
                    AuthSessionResult<AWSCognitoUserPoolTokens> authSessionResult =
                            cognitoAuthSession.getUserPoolTokensResult();
                    AWSCognitoUserPoolTokens userPoolTokens = authSessionResult.getValue();
                    String idToken = userPoolTokens.getIdToken();
                    assert idToken != null;
                    long auth_time = Long.parseLong(CognitoJWTParser.getClaim(idToken, "auth_time"));
                    String date = new java.text.SimpleDateFormat("dd/MM/yyyy HH:mm:ss").
                            format(new java.util.Date(auth_time * 1000));
                    Log.i("AuthQuickStart", "IdToken auth_time: " +  date + " " + auth_time);

                    long exp_time = Long.parseLong(CognitoJWTParser.getClaim(idToken, "exp"));
                    String exp_date = new java.text.SimpleDateFormat("dd/MM/yyyy HH:mm:ss").
                            format(new java.util.Date(exp_time * 1000));
                    Log.i("AuthQuickStart", "IdToken exp: " +  exp_date + " " + exp_time);
                },
                error -> Log.e("AuthQuickStart", error.toString())
        );
    }
```
Running this I get an expired token out of `fetchAuthSession` called with `forceRefresh(true)` :

```
2023-05-17 10:49:48.646 27220-27267 AuthQuickstart          com.example.myamplifyapp             I  Sign in succeeded
2023-05-17 10:49:48.986 27220-27267 AuthQuickStart          com.example.myamplifyapp             I  IdToken auth_time: 17/05/2023 10:49:46 1684313386
2023-05-17 10:49:48.987 27220-27267 AuthQuickStart          com.example.myamplifyapp             I  IdToken exp: 17/05/2023 10:56:46 1684313806
2023-05-17 10:57:49.404 27220-27268 AuthQuickStart          com.example.myamplifyapp             I  IdToken auth_time: 17/05/2023 10:49:46 1684313386
2023-05-17 10:57:49.406 27220-27268 AuthQuickStart          com.example.myamplifyapp             I  IdToken exp: 17/05/2023 10:56:46 1684313806
```

**Comment by tylerjroach** [comment]
Hi @Sypotu, 

I'm attempting to replicate the issue you are seeing but currently unable to. 

Can you please change the plugin order to:
```
Amplify.addPlugin(new AndroidLoggingPlugin(LogLevel.VERBOSE));
Amplify.addPlugin(new AWSCognitoAuthPlugin());
```

You will see much more logging for Auth related calls. Please post the logs (obfuscating where necessary) so we can analyze what is happening.

**Comment by Sypotu** [comment]
Hi @tylerjroach,

Thank you for your help, getting more logs helped me identify my problem, it was a configuration problem in my identity pool.

Maybe the documentation could be more clear about this order for the LoggingPlugin, I couldn't find it mentioned.

**Comment by tylerjroach** [comment]
@Sypotu Thank you for the suggestion. You are right that it isn't intuitive.

Would you be able to share the configuration issue you had in case others run into the same issue?

**Comment by aragabz** [comment]
> Hi @tylerjroach,
> 
> Thank you for your help, getting more logs helped me identify my problem, it was a configuration problem in my identity pool.
> 
> Maybe the documentation could be more clear about this order for the LoggingPlugin, I couldn't find it mentioned.

Would you like to share what was the issue? I am still facing this issue in my app

**Comment by ahardin** [comment]
I believe I found the problem (or one of them). You need to have `openid` in your list of Scopes in amplifyconfiguration.json. 

Note - this needs to be done at the time the initial access and refresh tokens are created. So, do a signout before testing this.

In hindsight, this should have been obvious, as the spec clearly indicates that `openid` is a required scope, and that "If the openid scope value is not present, the behavior is entirely unspecified." Unspecified indeed.

https://openid.net/specs/openid-connect-basic-1_0.html

cc: @aragabz @tylerjroach

**Comment by tylerjroach** [comment]
@ahardin Thank you! This is very helpful.

**Comment by tylerjroach** [comment]
Closing because of some reports of fixing through configuration issues. For anyone still experiencing an issue, please open a ticket with logs provided from the AmplifyLogging plugin.
